doc/user/inputs/blueprints.rst

```diff
@@ -404,9 +404,98 @@ material modifications
+                    ip: 0.0
+                    mult: 1
+                    op: 10.0
+        assemblies:
+            fuel a: &assembly_a
+                specifier: IC
+                blocks: [*block_fuel]
+                height: [1.0]
+                axial mesh points: [1]
+                xs types: [A]
+                material modifications:
+                    componentwise:
+                        fuel1:
+                            U235_wt_frac: [0.20]
+                        fuel2:
+                            Zr_wt_frac: [0.02]
+                    U235_wt_frac: [0.30]
+
+  This example would apply an enrichment of 20% to the ``fuel1`` component and an
+  enrichment of 30% to all other components in the block that accept the ``U235_wt_frac``
+  material modification. The user may alternatively use the ``blockwise`` attribute
+  to more clearly indicate that the material modification type is being applied to
+  all applicable components in the block::
+
+                material modifications:
+                    componentwise:
+                        fuel1:
+                            U235_wt_frac: [0.20]
+                        fuel2:
+                            Zr_wt_frac: [0.02]
+                    blockwise:
+                        U235_wt_frac: [0.30]
+
+  All componentwise material modifications override any blockwise material modifications
+  of the same type. In addition, any componentwise entries omitted for a given axial block
+  will default to the blockwise (or default, if no blockwise value is provided and a
+  default exists) value::
+
+        blocks:
+            fuel: &block_fuel
+                fuel1: &component_fuel_fuel1
+                    shape: Hexagon
+                    material: UZr
+                    Tinput: 600.0
+                    Thot: 600.0
+                    ip: 0.0
+                    mult: 1
+                    op: 10.0
+                fuel2: &component_fuel_fuel2
+                    shape: Hexagon
+                    material: UZr
+                    Tinput: 600.0
+                    Thot: 600.0
+                    ip: 0.0
+                    mult: 1
+                    op: 10.0
+        assemblies:
+            fuel a: &assembly_a
+                specifier: IC
+                blocks: [*block_fuel, *block_fuel]
+                height: [0.5, 0.5]
+                axial mesh points: [1, 1]
+                xs types: [A, A]
+                material modifications:
+                    componentwise:
+                        fuel1:
+                            U235_wt_frac: [0.20, ''] # <-- the U235_wt_frac for the second block will go to the blockwise value
+                        fuel2: # the U235_wt_frac for fuel2 component in both axial blocks will go to the blockwise value
+                            Zr_wt_frac: [0.02, ''] # <-- the Zr_wt_frac for the second block will default because there is no blockwise value
+                    blockwise:
```

Comment:
This answers my question above. This is nice that it keeps backwards compatibility. 

My main question here is - Is the length of the component and block mat mod lists required to be the same length as the number of blocks in the assembly? I believe they should be, but if not then we should make sure so there aren't any mistakes on the mapping. 

Also, it would be nice if there was a material modification summary table that could be produced for the user in the `stdout`. This is probably a separate diff, but would definitely help me verify that material mods are being applied as expected.

Reply:
Yeah, the lengths of the mat mods lists all have to be the same. If they are not, the code will crash (or once #515 is fixed, a good error will be printed).

Reply:
A summary table would definitely be useful. But I agree that it would be most appropriate in a separate PR.

Another thing that I think would be useful is a straight default mat mod value (not a list) on a given assembly so that the full list does not have to be written out in the case that all axial blocks take the same value. Obviously not needed, but could be convenient and reduce errors.

Reply:
That's true. It really does beg the question of what the mat mods are really supposed to do though. 

I understand that they interact with the material classes and there are some allowable kwargs for modifications, but it seems that with a full assembly modification we might get into the territory of considering setting number densities of nuclides directly on the blocks after the assembly had been modified. For example, reducing coolant density.

Maybe for now we just stick to what it is doing with applying mods to the material classes and also consider the full assembly mods and other uses for mat mods with @ntouran separately.

Reply:
> Yeah, the lengths of the mat mods lists all have to be the same. If they are not, the code will crash (or once #515 is fixed, a good error will be printed).

Do we have a unit test for that?  If not, can we get one?

Reply:
Unit tests are added for this function. I also went ahead and fixed #515 myself in addressing this, since the two were so tied up with each other.